src/irmin-containers/linked_log.ml

```diff
@@ -0,0 +1,230 @@
+end
+
+type ('a, 'b, 'c) store_item =
+  | Value of ('a, 'b, 'c) log_item
+  | Merge of ('a, 'b, 'c) log_item list
+
+module Store_item (T : Time.S) (K : Irmin.Hash.S) (V : Irmin.Type.S) :
+  Irmin.Type.S with type t = (T.t, V.t, K.t) store_item = struct
+  module L = Log_item (T) (K) (V)
+
+  type t = (T.t, V.t, K.t) store_item
+
+  let t =
+    let open Irmin.Type in
+    variant "t" (fun value merge ->
+      function Value v -> value v | Merge l -> merge l)
+    |~ case1 "Value" L.t (fun v -> Value v)
+    |~ case1 "Merge" (list L.t) (fun l -> Merge l)
+    |> sealv
+end
+
+module Log_ops
+    (C : Cas_maker.S)
+    (T : Time.S)
+    (K : Irmin.Hash.S)
+    (V : Irmin.Type.S) =
+struct
+  module Store = struct
+    include C.CAS_Maker (K) (Store_item (T) (K) (V))
+
+    let get_store =
+      let st = v @@ C.config in
+      fun () -> st
+  end
+
+  let append prev msg =
+    Store.get_store () >>= fun store ->
+    Store.batch store (fun t ->
+        Store.add t (Value { time = T.get_time (); msg; prev }))
+
+  let read_key k =
+    Store.get_store () >>= fun store ->
+    Store.find store k >>= function
+    | None -> failwith "not found"
+    | Some v -> return v
+
+  let sort l = List.sort (fun i1 i2 -> T.compare i2.time i1.time) l
+end
+
+module Linked_log
+    (C : Cas_maker.S)
+    (T : Time.S)
+    (K : Irmin.Hash.S)
+    (V : Irmin.Type.S) =
+struct
+  type t = K.t
+
+  let t = K.t
+
+  include Log_ops (C) (T) (K) (V)
+
+  let merge ~old:_ v1 v2 =
+    let open Irmin.Merge in
+    Store.get_store () >>= fun store ->
+    Store.find store v1 >>= fun v1 ->
+    Store.find store v2 >>= fun v2 ->
+    let lv1 =
+      match v1 with
+      | None -> []
+      | Some (Value v) -> [ v ]
+      | Some (Merge lv) -> lv
+    in
+    let lv2 =
+      match v2 with
+      | None -> []
+      | Some (Value v) -> [ v ]
+      | Some (Merge lv) -> lv
+    in
+    Store.batch store (fun t -> Store.add t (Merge (sort @@ lv1 @ lv2)))
+    >>= fun m -> ok m
+
+  let merge = Irmin.Merge.(option (v t merge))
+end
+
+module type S = sig
+  module Store : Irmin.S
+
+  type value
+
+  type cursor
+
+  val append :
+    info:Irmin.Info.f -> Store.t -> path:Store.key -> value -> unit Lwt.t
+
+  val get_cursor : Store.t -> path:Store.key -> cursor option Lwt.t
+
+  val read : cursor -> num_items:int -> (value list * cursor option) Lwt.t
+
+  val read_all : Store.t -> path:Store.key -> value list Lwt.t
+end
+
+module Make
+    (Backend : Irmin.S_MAKER)
+    (M : Irmin.Metadata.S)
+    (P : Irmin.Path.S)
+    (B : Irmin.Branch.S)
+    (H : Irmin.Hash.S)
+    (C : Cas_maker.S)
+    (T : Time.S)
+    (K : Irmin.Hash.S)
+    (V : Irmin.Type.S) : sig
+  include
+    S with type value = V.t and type Store.key = P.t and type Store.branch = B.t
+end = struct
+  module L = Linked_log (C) (T) (K) (V)
+  module Store = Backend (M) (L) (P) (B) (H)
+
+  module Set_elt = struct
+    type t = K.t
+
+    let compare h1 h2 = K.short_hash h1 - K.short_hash h2
+  end
```

Comment:
`Irmin.Hash.S.t` types are already comparable via their Irmin type. No need to define a new notion of comparison here.